plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/writer/GcsTransactionLogSynchronizer.java

```diff
@@ -0,0 +1,142 @@
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+
+        Storage storage = gcsStorageFactory.create(session);
+        PathCodec pathCodec = googleHadoopFileSystem.getGcsFs().getOptions().getPathCodec();
+        try {
+            createStorageObjectExclusively(newLogEntryPath.toUri(), entryContents, pathCodec, storage);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @Override
+    public boolean isUnsafe()
+    {
+        return false;
+    }
+
+    /**
+     * This method uses {@link Storage} Cloud Storage API client for creating synchronously a storage object
+     * with the specified path and content.
+     * <p>
+     * This approach avoids dealing with `FSDataOutputStream` exposed by the method
+     * `org.apache.hadoop.fs.FileSystem#create(org.apache.hadoop.fs.Path)` and the intricacies of handling
+     * exceptions which may occur while writing the content to the output stream.
+     */
+    private static StorageObject createStorageObjectExclusively(URI blobPath, byte[] content, PathCodec pathCodec, Storage storage)
+            throws IOException
+    {
+        CreateFileOptions createFileOptions = new CreateFileOptions(false);
+        CreateObjectOptions createObjectOptions = objectOptionsFromFileOptions(createFileOptions);
+        StorageResourceId storageResourceId = pathCodec.validatePathAndGetId(blobPath, false);
+
+        StorageObject object =
+                new StorageObject()
+                        .setContentEncoding(createObjectOptions.getContentEncoding())
+                        .setMetadata(encodeMetadata(createObjectOptions.getMetadata()))
+                        .setName(storageResourceId.getObjectName());
+
+        InputStream inputStream = new ByteArrayInputStream(content, 0, content.length);
+        Storage.Objects.Insert insert = storage.objects().insert(
+                storageResourceId.getBucketName(),
+                object,
+                new InputStreamContent(createObjectOptions.getContentType(), inputStream));
+        // By setting `ifGenerationMatch` setting to `0`, the creation of the blob will succeed only
+        // if there are no live versions of the object. When the blob already exists, the operation
+        // will fail with the exception message `412 Precondition Failed`
+        insert.setIfGenerationMatch(0L);
+        insert.getMediaHttpUploader().setDirectUploadEnabled(true);
+        insert.setName(storageResourceId.getObjectName());
+        return insert.execute();
+    }
+
+    /**
+     * Helper for converting from a {@code Map<String, byte[]>} metadata map that may be in a
+     * StorageObject into a {@code Map<String, String>} suitable for placement inside a
+     * {@link GoogleCloudStorageItemInfo}.
+     */
+    private static Map<String, String> encodeMetadata(Map<String, byte[]> metadata)
```

Comment:
Haven't actually looked myself, but this feels like something that should exist in the library already

Reply:
The reason for this code duplication is that the method `com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl#encodeMetadataValues` is actually `private static` and therefore not accessible for reuse in this  code.